x/programs/wasmlanche/src/memory.rs

```diff
@@ -13,22 +13,106 @@ use alloc::{
     alloc::{alloc as allocate, dealloc as deallocate, handle_alloc_error, Layout},
     vec::Vec,
 };
-use core::{
-    cell::{LazyCell, RefCell},
-    mem::ManuallyDrop,
-    ops::Deref,
-    slice,
-};
-use hashbrown::HashMap;
+use core::{mem::ManuallyDrop, ops::Deref, slice};
+
+/// Safety:
+/// Only safe to use in a single-threaded environment
+mod allocations {
+    use cfg_if::cfg_if;
+    use core::cell::RefCell;
+    use hashbrown::HashMap;
+
+    type LenMap = HashMap<*const u8, usize>;
+
+    // Map of pointer to the length of its content on the heap
+    cfg_if! {
+        if #[cfg(feature = "test")] {
+            std::thread_local! {
+                static ALLOCATIONS: RefCell<LenMap> = RefCell::new(HashMap::new());
+            }
+        } else {
+            extern crate alloc;
+
+            use alloc::boxed::Box;
+            use core::sync::atomic::{AtomicPtr, AtomicBool, Ordering::{Acquire, Relaxed, Release}};
+
+            static ALLOCATIONS: SingletonLenMap = SingletonLenMap::new();
+            static LOCK: AtomicBool = AtomicBool::new(false);
+
+            struct Guard;
+
+            impl Drop for Guard {
+                fn drop(&mut self) {
+                    LOCK.store(false, Release);
+                }
+            }
+
+            struct SingletonLenMap(AtomicPtr<RefCell<LenMap>>);
+
+            // Safety: Can only initialize once on one thread.
+            unsafe impl Sync for SingletonLenMap {}
+
+
+            impl SingletonLenMap {
+                const fn new() -> Self {
+                    Self(AtomicPtr::new(core::ptr::null_mut()))
+                }
+
+                #[must_use]
+                fn init(&self) -> Guard {
```

Comment:
function name was a bit misleading, seemed like it was initing the `SingletonLenMap`. Maybe call it `acquire_guard` or `init_guard`

Reply:
ah actually this is doing two things; both grabbing the guard and initing the map if not created. Still think renaming would make this clearer(`init_gaurd` doesn't make sense anymore but `acquire_guard` could.